diff-test/core/src/main.ts

```diff
@@ -0,0 +1,437 @@
+};
+
+type ConsumerSlash = {
+  kind: string;
+  val: number;
+  infractionHeight: number;
+  isDowntime: number;
+};
+
+/**
+ * Takes an object where values are probabilities and returns a random
+ * key based on the distribution.
+ */
+function weightedRandomKey(distr) {
+  const scalar = _.reduce(_.values(distr), (sum, y) => sum + y, 0);
+  const x = Math.random() * scalar;
+  const pairs = _.pairs(distr);
+  let i = 0;
+  let cum = 0;
+  while (i < pairs.length - 1 && cum + pairs[i][1] < x) {
+    cum += pairs[i][1];
+    i += 1;
+  }
+  return pairs[i][0];
+}
+
+class ActionGenerator {
+  model;
+  didSlash = new Array(NUM_VALIDATORS).fill(false);
+
+  constructor(model) {
+    this.model = model;
+  }
+
+  /**
+   * Get a new model action.
+   * An action is chosen by a process of generating and then selecting
+   * template actions based on the current state of the model.
+   * In this way, only actions which make sense given the current state
+   * of the model can be returned. Among the possible actions, one is
+   * chosen based on a probability distribution.
+   * @returns An executable model action
+   */
+  get = () => {
+    // Get candidate actions
+    let templates: Action[] = _.flatten([
+      this.candidateDelegate(),
+      this.candidateUndelegate(),
+      this.candidateJumpNBlocks(),
+      this.candidateDeliver(),
+      this.candidateConsumerSlash(),
+    ]);
+    // Get the names of each possible action
+    const possibleActions = _.uniq(templates.map((a) => a.kind));
+    // Build a probability distribution for each possible action
+    const distr = _.pick(
+      {
+        Delegate: 0.03,
+        Undelegate: 0.03,
+        JumpNBlocks: 0.37,
+        Deliver: 0.55,
+        ConsumerSlash: 0.02,
+      },
+      ...possibleActions,
+    );
+    // Choose an action type (kind) based on the probability distribution
+    const kind = weightedRandomKey(distr);
+    templates = templates.filter((a) => a.kind === kind);
+    const a = _.sample(templates);
+    if (kind === 'Delegate') {
+      return this.selectDelegate(a);
+    }
+    if (kind === 'Undelegate') {
+      return this.selectUndelegate(a);
+    }
+    if (kind === 'JumpNBlocks') {
+      return this.selectJumpNBlocks(a);
+    }
+    if (kind === 'Deliver') {
+      return this.selectDeliver(a);
+    }
+    if (kind === 'ConsumerSlash') {
+      return this.selectConsumerSlash(a);
+    }
+    throw `kind doesn't match`;
+  };
+
+  // Return templates for possible delegate actions
+  candidateDelegate = (): Action[] => {
+    return _.range(NUM_VALIDATORS).map((i) => {
+      return {
+        kind: 'Delegate',
+        val: i,
+      };
+    });
+  };
+
+  // Fill out template for a selected Delegate action
+  selectDelegate = (a): Delegate => {
+    return { ...a, amt: _.random(1, 5) * TOKEN_SCALAR };
+  };
+
+  // Return templates for possible undelegate actions
+  candidateUndelegate = (): Action[] => {
+    return _.range(NUM_VALIDATORS).map((i) => {
```

Comment:
[...Array(NUM_VALIDATORS).keys()].map((i) => {...})

Reply:
That's not an improvement IMO